tests/subsys/bluetooth/services/ble_bas/src/unity_test.c

```diff
@@ -0,0 +1,450 @@
+		TEST_ASSERT_TRUE(p_char_md->char_props.notify);
+	}
+	TEST_ASSERT_TRUE(p_char_md->char_props.read);
+
+	TEST_ASSERT_NOT_NULL(p_attr_char_value);
+	TEST_ASSERT_NOT_NULL(p_attr_char_value->p_attr_md);
+	TEST_ASSERT_EQUAL(BLE_GATTS_VLOC_STACK, p_attr_char_value->p_attr_md->vloc);
+	TEST_ASSERT_EQUAL(perm_12.lv, p_attr_char_value->p_attr_md->read_perm.lv);
+	TEST_ASSERT_EQUAL(perm_12.sm, p_attr_char_value->p_attr_md->read_perm.sm);
+	TEST_ASSERT_EQUAL(expected_bat_lvl, *p_attr_char_value->p_value);
+
+	TEST_ASSERT_NOT_NULL(p_handles);
+	p_handles->value_handle = VALUE_HANDLE;
+	p_handles->user_desc_handle = USER_DESC_HANDLE;
+	p_handles->cccd_handle = CCCD_HANDLE;
+	p_handles->sccd_handle = SCCD_HANDLE;
+
+	return NRF_SUCCESS;
+}
+
+static uint32_t stub_sd_ble_gatts_descriptor_add_success(uint16_t char_handle,
+							 const ble_gatts_attr_t *p_attr,
+							 uint16_t *p_handle, int calls)
+{
+	const ble_gap_conn_sec_mode_t perm_56 = {.lv = 5, .sm = 6};
+	const struct {
+		uint8_t report_id;
+		uint8_t report_type;
+	} expected_report_ref = { .report_id = 1, .report_type = 0x01 };
+
+	TEST_ASSERT_EQUAL(VALUE_HANDLE, char_handle);
+
+	TEST_ASSERT_NOT_NULL(p_attr);
+	TEST_ASSERT_NOT_NULL(p_attr->p_uuid);
+	TEST_ASSERT_EQUAL(BLE_UUID_TYPE_BLE, p_attr->p_uuid->type);
+	TEST_ASSERT_EQUAL(BLE_UUID_REPORT_REF_DESCR, p_attr->p_uuid->uuid);
+	TEST_ASSERT_NOT_NULL(p_attr->p_attr_md);
+	TEST_ASSERT_EQUAL(BLE_GATTS_VLOC_STACK, p_attr->p_attr_md->vloc);
+	TEST_ASSERT_EQUAL(perm_56.lv, p_attr->p_attr_md->read_perm.lv);
+	TEST_ASSERT_EQUAL(perm_56.sm, p_attr->p_attr_md->read_perm.sm);
+
+	TEST_ASSERT_EQUAL(expected_report_ref.report_id, p_attr->p_value[0]);
+	TEST_ASSERT_EQUAL(expected_report_ref.report_type, p_attr->p_value[1]);
+
+	TEST_ASSERT_NOT_NULL(p_handle);
+	*p_handle = REPORT_REF_HANDLE;
+
+	return NRF_SUCCESS;
+}
+
+static void ble_bas_evt_handler(struct ble_bas *bas, const struct ble_bas_evt *evt)
+{
+	evt_handler_called = true;
+}
+
+static void ble_bas_evt_handler_notif_enabled(struct ble_bas *bas, const struct ble_bas_evt *evt)
+{
+	TEST_ASSERT_EQUAL(BLE_BAS_EVT_NOTIFICATION_ENABLED, evt->evt_type);
+	evt_handler_called = true;
+}
+
+static void ble_bas_evt_handler_notif_disable(struct ble_bas *bas, const struct ble_bas_evt *evt)
+{
+	TEST_ASSERT_EQUAL(BLE_BAS_EVT_NOTIFICATION_DISABLED, evt->evt_type);
+	evt_handler_called = true;
+}
+
+void bas_init(const struct ble_bas_config *cfg)
+{
+	int ret;
+
+	__cmock_sd_ble_gatts_service_add_Stub(stub_sd_ble_gatts_service_add_success);
+	__cmock_sd_ble_gatts_characteristic_add_Stub(stub_sd_ble_gatts_characteristic_add_success);
+	__cmock_sd_ble_gatts_descriptor_add_Stub(stub_sd_ble_gatts_descriptor_add_success);
+
+	ret = ble_bas_init(&ble_bas, cfg);
+	TEST_ASSERT_EQUAL(0, ret);
+	TEST_ASSERT_EQUAL(SERVICE_HANDLE, ble_bas.service_handle);
+	TEST_ASSERT_EQUAL(VALUE_HANDLE, ble_bas.battery_level_handles.value_handle);
+	TEST_ASSERT_EQUAL(REPORT_REF_HANDLE, ble_bas.report_ref_handle);
```

Comment:
You can't test these like that-- but you can try to call some API that would pass these values as a parameter to a SoftDevice call to verify that they were set as expected during `ble_bas_init()`

Reply:
For example `ble_bas_battery_level_notify()` passes the field `battery_level_handles.value_handle` to the `sd_ble_gatts_hvx()` function